File: backend/app/schemas/analytics.py
from pydantic import BaseModel
from typing import List, Dict, Any, Optional
from datetime import datetime


class FieldActivityAnalytics(BaseModel):
    total_activities: int
    total_hours: float
    activities_this_week: int
    activities_this_month: int
    top_staff: List[Dict[str, Any]] = []  # Top field staff by activity count
    category_distribution: Dict[str, int] = {}  # Activities by category
    top_customers: List[Dict[str, Any]] = []  # Top customers by visit count
    unique_customers: int = 0  # Number of unique customers served
    # Billing metrics
    billable_hours: float = 0.0  # Total hours for ON_SITE visits
    non_billable_hours: float = 0.0  # Total hours for OFFICE activities
    billable_visits: int = 0  # Count of ON_SITE activities
    office_activities: int = 0  # Count of OFFICE activities
    billing_rate: float = 0.0  # Percentage of billable hours


class TaskAnalytics(BaseModel):
    total_tasks: int
    completed_tasks: int
    in_progress_tasks: int
    overdue_tasks: int
    completion_rate: float
    average_completion_time: Optional[float] = None  # in days


class ProjectAnalytics(BaseModel):
    project_id: int
    project_name: str
    total_tasks: int
    completed_tasks: int
    completion_rate: float

    class Config:
        from_attributes = True


class UserProductivity(BaseModel):
    user_id: int
    user_name: str
    tasks_assigned: int
    tasks_completed: int
    completion_rate: float
    average_hours_per_task: Optional[float] = None


class WorkspaceAnalytics(BaseModel):
    workspace_id: int
    workspace_name: str
    total_projects: int
    total_tasks: int
    completed_tasks: int
    active_members: int
    completion_rate: float
    projects: List[ProjectAnalytics] = []
    field_activities: Optional[FieldActivityAnalytics] = None  # Field operations data


class ExecutiveDashboard(BaseModel):
    overview: TaskAnalytics
    workspaces: List[WorkspaceAnalytics] = []
    top_performers: List[UserProductivity] = []
    task_completion_trend: List[Dict[str, Any]] = []  # Time series data
    priority_distribution: Dict[str, int] = {}
    status_distribution: Dict[str, int] = {}

File: frontend/src/types/analytics.ts
export interface FieldActivityAnalytics {
  total_activities: number;
  total_hours: number;
  activities_this_week: number;
  activities_this_month: number;
  top_staff: Array<{
    user_id: number;
    name: string;
    activity_count: number;
  }>;
  category_distribution: Record<string, number>;
}

export interface TaskAnalytics {
  total_tasks: number;
  completed_tasks: number;
  in_progress_tasks: number;
  overdue_tasks: number;
  completion_rate: number;
  average_completion_time: number | null;
}

export interface ProjectAnalytics {
  project_id: number;
  project_name: string;
  total_tasks: number;
  completed_tasks: number;
  completion_rate: number;
}

export interface UserProductivity {
  user_id: number;
  user_name: string;
  tasks_assigned: number;
  tasks_completed: number;
  completion_rate: number;
  average_hours_per_task: number | null;
}

export interface WorkspaceAnalytics {
  workspace_id: number;
  workspace_name: string;
  total_projects: number;
  total_tasks: number;
  completed_tasks: number;
  active_members: number;
  completion_rate: number;
  projects: ProjectAnalytics[];
  field_activities?: FieldActivityAnalytics | null;
}

export interface ExecutiveDashboard {
  overview: TaskAnalytics;
  workspaces: WorkspaceAnalytics[];
  top_performers: UserProductivity[];
  task_completion_trend: Array<{
    date: string;
    completed: number;
  }>;
  priority_distribution: Record<string, number>;
  status_distribution: Record<string, number>;
}
